lmdeploy/pytorch/kernels/dipu/apply_rotary_pos_emb.py

```diff
@@ -0,0 +1,32 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import deeplink_ext.cpp_extensions as ext
+import torch
+from torch import Tensor
+
+
+def apply_rotary_pos_emb(
+    query_states: Tensor,
+    key_states: Tensor,
+    cos: Tensor,
+    sin: Tensor,
+    position_ids: Tensor,
+    position_ids_1d: Tensor,
+    q_embed=None,
+    k_embed=None,
+    context=None,
+):
+    bs, head, dim = query_states.shape
+    numKeyValueHeads = key_states.shape[1]
+    seqlens = [(min(position_id), max(position_id) + 1)
+               for position_id in position_ids.tolist()]
+    query_states = query_states.reshape(bs, head * dim)
+    key_states = key_states.reshape(bs, numKeyValueHeads * dim)
+    if not (hasattr(context, 'cos') or hasattr(context, 'sin')):
+        cos = torch.cat([cos[i:j] for i, j in seqlens]).view(1, bs, 1, -1)
```

Comment:
Can we index it with position_ids_1d?

Reply:
I understand what you mean, I will modify it in the next commit.